### ISSUE TITLE
tab widget

### ISSUE THREAD

**Issue by Xeverous** [cycfi/elements#23]
I would like to implement tab (and some other widgets) but I need some guidance regarding library's code - everything is so composable that I have a feeling I will write duplicated code most of time time.

I have read [the post](https://www.cycfi.com/2019/07/photon-micro-gui/) and checked how button is implemented (array composite of 2 elements) but I still don't get many things:

- What is the purpose of `_base` types? That is, `proxy` vs `proxy_base`, `composite` vs `composite_base` etc
- Why do various elements are class templates and inherit from `typename Base`? What's the benefit of doing this over just inheriting from `element`?
  - Are there any restrictions on the inheritance (eg `Base` must be `element` or a type derived from it)?
- I figured out that the `support` directory is a collection of helpers (implementation details etc) but what is the reliation of `element/*` to `element/gallery/*`? Both directory trees contain `button` header and source - their reliation/dependency is unclear to me.
- What are some very good widget example implementations that I should look at when implementing other widgets (eg if I would like to implement radio box, I should surely check check box)?

**Comment by djowel** [comment]
On one hand, I was thinking there's a lot I need to answer here and for the sake of expediency, I thought about doing a video chat on this, and your other post on radio-boxes. On the other hand, it might also be good to write down my answer that can become basis for documentation. I'm a bit torn on how to proceed, given that I am also in the middle of something else. Allow me some time to think about how best to approach this, OK? A day or two perhaps...

**Comment by Xeverous** [comment]
Ok, I'm fine with realtime conversation. From my previous experience from work and other collaborations (mostly university stuff) the most productive way to solve such problems was to actually do all of these things together, that is:

- realtime voice chat (with screen sharing if some code/schemas/diagrams was needed to be presented)
- writing documentation basis during the conversation
- writing down any agreements/decisions (usually things to do for everyone up to the next meeting)

Looking forward to your decision.

**Comment by djowel** [comment]
Sounds good to me. Actually, I'm thinking the same... Let's do it. Email me at djowel-at-gmail-dot-com and let's schedule a video chat.

**Comment by Xeverous** [comment]
So the design has been quite well-defined during our latest talk, but I have some implementation-related questions:

For now, lets assume the widget is named `notebook` as it is the best name for me currently and you haven't commended on this.

- base class `notebook_element_base`
- classes `vnotebook_element`, `hnotebook_element`
- factory functions `vnotebook`, `hnotebook` (each with 2 overloads) that take 2 arguments (bar of tabs + the screened deck element) -  This will cover all 4 layouts where the bar of tabs has different placement (top, bottom, left, right). Then such factory functions call `vtile`/`htile` to create a single widget that automatically implements functions such as limits, events etc.

Example use would be:

```cpp
// +------+------+------+------+-----+
// | tab0 | tab1 | tab2 | tab3 | ... |
// +------+------+------+------+-----+
// |                                 |
// |            screen               |
// |         (deck element)          |
// |                                 |
// |                                 |
// +---------------------------------+

auto w = vnotebook(
    htile(tab0, tab1, tab2, ...), // bar of tabs
    deck(screen0, screen1, screen2, ...) // deck element which switches depending on tab
);
```

My questions:

- What the argument types for the notebook factory functions should be? Currently (from what `vtile`/`htile` and `deck` factory function returns) `notebook` factory functions should accept `array_composite<N, v/htile_element>` and `array_composite<M, deck_element>` and `static_assert(N == M)`.
- Deck element (what I call the notebook screen) switching depending on the selected tab is very trivial to implement but in order to make tabs clickable, each of them should be wrapped in a button - am I right? I don't see a clean way to perform this if the factory function would accept a composite which already has an array of v/htiles. Do we want a `key_intercept` there? What's your view on this problem?
- I'm concerned how the bar of tabs would be rendered - so far it consists of few elements that do not change appearance. Should we move this burden to the user (and expect the user to pass a proper "bar of tabs") or should the notebook factory (apart from making tabs clickable) wrap the bar of tabs in some shadows to signal which one is selected?

Unrelated question:

- Why does each htile/vtile element have a vector? Why each element in a composite needs this vector?

**Comment by djowel** [comment]
This will be a little bit terse, but I'll see if I can reply with more substance tomorrow.

For now, I suggest working on the `tab` (very similar to `radio_button`) and refactoring the similarities. You'll have to do some simple `canvas` drawings for the tab image (un-selected and selected  --typically the selected is lighter than the unselected), using this as the basis:

https://github.com/cycfi/elements/blob/develop/lib/src/element/gallery/radio_button.cpp#L10

where:
1. state: selected or unselected
2. hilite: mouse overs over the button (showing some feedback that it is clickable)

Then you can make an example using it, again following the radio_button example in the buttons example.

Let's continue when you have a working tab group example (htile of `tab`s).

**Comment by Xeverous** [comment]
`radio_button::select` has some logic duplication (state is checked twice):

```cpp
   void basic_radio_button::select(bool state)
   {
      if (state != is_selected())
         value(state);
   }

   bool basic_radio_button::is_selected() const
   {
      return value();
   }

   bool layered_button::value() const
   {
      return _state;
   }

   void layered_button::value(bool new_state)
   {
      if (_state != new_state)
         state(new_state);
   }
```

It could be just:

```cpp
   void basic_radio_button::select(bool state)
   {
      value(state);
   }
```

**Comment by Xeverous** [comment]
Trying to grasp and move upwards the similarities between `check_box` and `radio_button` - I don't like the situation that both widgets create copies of the string and actually store 2x (string + state representation widget) instead of 2x state representation widget + 1x string.

**Comment by Xeverous** [comment]
I think that radio buttons and checkboxes should be using `label` factory. Then someone could easily `htile` them to decide whether they want text on the left or right side of the check/radio.

**Comment by djowel** [comment]
> It could be just:
> 
> ```c++
>    void basic_radio_button::select(bool state)
>    {
>       value(state);
> ```

I think you assuming that `is_selected()` and `value()` are `final` (which is probably a good assumption).

**Comment by djowel** [comment]
> Trying to grasp and move upwards the similarities between `check_box` and `radio_button` - I don't like the situation that both widgets create copies of the string and actually store 2x (string + state representation widget) instead of 2x state representation widget + 1x string.

Yes, I thought so too. That is also why I started the text_reader thing: https://github.com/cycfi/elements/blob/develop/lib/include/elements/element/text.hpp#L23

**Comment by Xeverous** [comment]
Possible, that migth simply be an "overassumption" considering there are many virtual calls.

**Comment by Xeverous** [comment]
So, got any rought plans how a checkbox/radio button would store text? How would you compose them using label and `text_reader` interface?

**Comment by djowel** [comment]
> So, got any rought plans how a checkbox/radio button would store text? How would you compose them using label and `text_reader` interface?

The basic idea is that one should be able to make a label that uses a string_view, (-- even an entirely generated string, but that is not relevant here). I'm not sure about the API yet (see below). Let's cross that bridge later :-)

For example, I would like to detect or have the user hint on literal strings. In such cases, you don't even need a std::string. And there are many cases where you just want to pass string_view. The string is already stored somewhere else (e.g. some resources for internationalization e.g. a std::map)... Basically a lifetime management issue.

**Comment by Xeverous** [comment]
You can make multiple constructors:

- `const char(&)[N]` - this must be a C-string, guuaranteed infinite lifetime
- `std::string_view` - user manages lifetime
- `std::string` - widget manages lifetime
- `const char* = delete` because it is unknown, require the user to wrap the call into `string` or `string_view`.

**Comment by djowel** [comment]
> > So, got any rought plans how a checkbox/radio button would store text? How would you compose them using label and `text_reader` interface?
> 
> The basic idea is that one should be able to make a label that uses a string_view, (-- even an entirely generated string, but that is not relevant here). I'm not sure about the API yet (see below). Let's cross that bridge later :-)
> 
> For example, I would like to detect or have the user hint on literal strings. In such cases, you don't even need a std::string. And there are many cases where you just want to pass string_view. The string is already stored somewhere else (e.g. some resources for internationalization e.g. a std::map)... Basically a lifetime management issue.

In other words, the source of the text/string should not matter for things like labels. And we need to give the client ways to hint at the source of the string which determines if it will be stored or not.

**Comment by djowel** [comment]
> You can make multiple constructors:
> 
> * `const char(&)[N]` - this must be a C-string, guuaranteed infinite lifetime
> * `std::string_view` - user manages lifetime
> * `std::string` - widget manages lifetime
> * `const char* = delete` because it is unknown, require the user to wrap the call into `string` or `string_view`.

Exactly! But not on the ctor, instead the factory...

**Comment by Xeverous** [comment]
To summarize the current state:

- notebook needs tabbar
- tabbar wants (not must-have but better refactor now) improved `radio_button` and `check_box`
- these widgets or the `label` needs to implement `text_reader` interface

What is the current state of `text_reader`? Are you working on it? I would like to move forward with the tabbar but don't want any conflicts.

**Comment by djowel** [comment]
> What is the current state of `text_reader`? Are you working on it? I would like to move forward with the tabbar but don't want any conflicts.

Please proceed. I'm working n something else.

**Comment by Xeverous** [comment]
I'm back and ready for more contributions. I see there have been made some changes by @dmacvicar  (CMake, CI, GTK)

I can confirm that both `master` and `develop` branches build for me, haven't tried GTK on Windows yet.

- Is removing infra still planned? I think I could easily edit all assertions and min/max/clamp calls.
- What is the current state of text rendering? I want to push tab widget implementation. Has nothing changed since my last reply?
- Is this actually possible (any permutation of cocoa, GTK, win32 on any of WIN/UNIX/APPLE)? IMO the chache strings should be edited to actually represent what is possible on each system.

https://github.com/cycfi/elements/blob/ccb2f2c4107827cb9f9d98e8fb3daf3944e0567f/CMakeLists.txt#L30-L38

**Comment by djowel** [comment]
> * Is removing infra still planned? I think I could easily edit all assertions and min/max/clamp calls.

No, actually the filesystem namespace workaround moved to it since we removed the boost::filesystem dependency. But this time, infra is a submodule with header only dependencies and the link problems with other boost libs have been fixed (see my commit remarks).

> * What is the current state of text rendering? I want to push tab widget implementation. Has nothing changed since my last reply?

Text rendering is good. I am busy working on "artist": the "canvas" part of Elements made modular and supporting other backends such as Skia, Quartz2D and Direct2D.

> * Is this actually possible (any permutation of cocoa, GTK, win32 on any of WIN/UNIX/APPLE)? IMO the chache strings should be edited to actually represent what is possible on each system.

Very good point. I'd love to see what you can make out of it. I'll probably need something like it in "Artist" with it's different backends.

**Comment by Xeverous** [comment]
How about removing duplicate string instances in buttons? Various widgets like radio button and check box are creating multiple layered elements in order to display their on/off states but IMO these layers should only contain the icon, string should be just one. If this is not changed before I implement notebook then it will be another thing to refactor and since notebook is more complex than these I'm afraid it will require more effort to change.

**Comment by djowel** [comment]
> How about removing duplicate string instances in buttons? Various widgets like radio button and check box are creating multiple layered elements in order to display their on/off states but IMO these layers should only contain the icon, string should be just one. If this is not changed before I implement notebook then it will be another thing to refactor and since notebook is more complex than these I'm afraid it will require more effort to change.

Ah right. Yes, thanks for reminding me. Would you open an issue for this? It's low prio for now, but we should have some facilities now to do this. Ideally, you should not even have to hold strings if in many cases they are constant literals. The basic issue is how strings are held and how to detect ownership.

**Comment by Xeverous** [comment]
Opened #87 and #88. Got any information that could help with working on #87?

**Comment by djowel** [comment]
> Opened #87 and #88. Got any information that could help with working on #87?

Sure! I'll post some comments on the issue. Thanks for doing that BTW!

**Comment by Xeverous** [comment]
My current "starter" implementation, based on `v/htile` and `composite`:

```cpp
	class notebook_element_base
	{
	
	};
	
	class vnotebook_element : public notebook_element_base, public composite_base
	{
	
	};
	
	template <std::size_t N, std::size_t M>
	auto vnotebook(
		array_composite<N, htile_element> tabs,
		array_composite<M, deck_element> screens)
	{
		static_assert(N == M);
		using composite = array_composite<2, vnotebook_element>;
		using container = typename composite::container_type;
		composite r{};
		r = container{{ share(std::move(tabs)), share(std::move(screens)) }};
		return r;
	}

	class hnotebook_element : public notebook_element_base, public composite_base
	{
	
	};

	template <std::size_t N, std::size_t M>
	auto hnotebook(
		array_composite<N, vtile_element> tabs,
		array_composite<M, deck_element> screens)
	{
		static_assert(N == M);
		using composite = array_composite<2, hnotebook_element>;
		using container = typename composite::container_type;
		composite r{};
		r = container{{ share(std::move(tabs)), share(std::move(screens)) }};
		return r;
	}
```

I have no idea how to progress. I don't get why each `vtile_element` has a vector of `float`s and why the actual composite needs to inherit from both container and `vtile_element`. I thought that a composite would store such `vtile_element`s, not inherit from such type. I can't get how to implement notebook functions, reading v/htile implementation does not help. The only thing I can state now (in regards to design) that notebook should stretch only the screened deck and allocate minimal room for the bar of tabs.

**Comment by Xeverous** [comment]
I made a sample tab implementation, see 652e32e87f57a6dc36a5b2f13d10653962dc429a in my fork

**Comment by Xeverous** [comment]
The rendering is obviously far from being correct but you get the point ... as was expected the implementation is an analogic copy of the radio button (only drawing differs). The question is where the duplicated code should be extracted - `selectable`?

**Comment by djowel** [comment]
<img width="651" alt="Screen Shot 2020-05-06 at 8 13 27 PM" src="https://user-images.githubusercontent.com/568094/81175650-76f89f80-8fd6-11ea-94a6-e99c25fda6a9.png">

There. It is working. I added an basic example for it. Check it out. Let's see how you are able to capture that in a gallery API. Pushed in experimental branch.

**Comment by djowel** [comment]
Ooops. Compile errors on gcc/msvc. See if you can fix it. Otherwise, I'll do it tomorrow.

**Comment by djowel** [comment]
OK, gcc and clang builds now. It's down to silly msvc. probably a missing include...

**Comment by Xeverous** [comment]
Yay! It works. I noticed only some warnings, I can fix them.

Now obviously it would be useful to have a notebook factory that frees the library user from writing a function like the `make_tabs` in the example. But I also see that the retuned `vtile` is arbitrarily nested so the widget would need to use something like `find_composite`, am I right?

**Comment by Xeverous** [comment]
The click result (`r`) is not used. Unneeded or something was forgotten?

https://github.com/cycfi/elements/blob/4492395705c074ef08b64f77bc94c9d2332e4668/lib/include/elements/element/button.hpp#L191-L205

**Comment by djowel** [comment]
I just worked on it as fast as I can. So there should be lots of cleaning up to do. I'll go check tomorrow. I'm done for today. Now your challenge is to make that an easy-to-use gallery API :-)

**Comment by djowel** [comment]
please pull latest. i just fixed an msvc thing.

**Comment by Xeverous** [comment]
I fixed the warnings.

I don't get this commit (cf2ae3afa154a7c65e43d6dd7c8c9f587c4258bd) - if there is a need for `if constexpr` depending on the inheritance then it could potentially be a compiler-intepended issue. But the include replaced with forward declaration (circular dependency)? Is this a separate problem? I don't get what did you fixed there.

**Comment by djowel** [comment]
> I don't get this commit ([cf2ae3a](https://github.com/cycfi/elements/commit/cf2ae3afa154a7c65e43d6dd7c8c9f587c4258bd)) - if there is a need for `if constexpr` depending on the inheritance then it could potentially be a compiler-intepended issue. But the include replaced with forward declaration (circular dependency)? Is this a separate problem? I don't get what did you fixed there.

No. It looks like a compiler bug. I'll see if there's a better fix tomorrow.

**Comment by Xeverous** [comment]
Coming back to this issue. To summarize and state what is needed:

- We have a `tab` element implementation that works very similarly to `radio_button`.
- We have a `deck` element that can be used to represent notebook's main screen
- The notebook itself does not require any extra element implementation, all that is needed is tiling screen + tabs together and handlers on tab selection. The first one is already covered by tile factory functions.
- The second one - automatic implementation of tab switching is not yet present and will not be trivial to implement. Right now, the example does it by itself by explictly writing and assigning all handlers. Of course I would like elements to take care of it automatically but there is a problem: tabs and the deck element can be arbitrarily nested. This makes typical factory function composer apporach impossible.

The thing I can think of right now is a apply-like functions that takes `(deck&, tab&...)` instead, assigns all handlers and returns nothing. User would need to explicitly provide all widgets that have some external lifetime.

**Comment by djowel** [comment]
> * tabs and the deck element can be arbitrarily nested. This makes typical factory function composer apporach impossible.

Never say impossible :-) There is a way. Hint: walk the trees!

**Comment by djowel** [comment]
> Hint: walk the trees!

Oh and by doing that, I probably want a generic set of traversals encapsulated in a group of algorithms, similar to find_composite, find_proxy and find_element. These need to be fleshed out and made generic.

**Comment by Xeverous** [comment]
I have thought about such solution but it didn't appeal to me:

- I have limited knowledge of internals how elements are traversed (you will be better at your own implementation)
- I'm concerned with so pervasive use of RTTI. I don't have a need to make a non-RTTI build, but it concerns me about performance. Elements samples take 10-20% CPU on a 3-4GHz CPU when just moving mouse around widgets and I wonder what's causing such high processor usage. Obviously there is no good argument without measurement but I'm thinking elements could get some good CPU => RAM usage tradeoffs. Have you though about some way of caching computed values when widgets do not need refreshing?

**Comment by Xeverous** [comment]
One more thing: if you continue to implement things on this branch, please rebase it before merging. I often need to check something in elements history and having unchronological merge commits does not make it easier.

**Comment by djowel** [comment]
> * I'm concerned with so pervasive use of RTTI. I don't have a need to make a non-RTTI build, but it concerns me about performance.

Definitely not in this case! This only happens at construction time. And yes of course, as always, when in doubt, don't guess. Measure. I'm pretty sure optimization will be one to deal with, but not at this point. Premature optimization is never a good thing. 

Also, there are other techniques beyond RTTI such as static tag dispatch. But I doubt that's the perf issue you are alluding too. Whatever it is, it has to be measured and optimized. I might have an idea, but I do not want to guess.

Yes, I have a few optimization ideas already in mind.

**Comment by Xeverous** [comment]
Can you implement "walking the tree"? I'm not very knowledgable on `find_composite`, `find_proxy`, `find_element` etc and tabs are my highest-wanted feature since a long time. This feature requires a lot of internal implementation knowledge that only you have. I think you could easily add a function like `vnotebook(element& deck, element& tabs)` that walks the hierarchy downward and applies handlers on certain elements.

**Comment by djowel** [comment]
@Xeverous, typically, in the past OSS projects I authored (e.g. Boost Spirit, Fusion, and Phoenix), collaborators go figure out by themselves how things work internally (there were no internal docs).   And the internals of Spirit, Fusion, Phoenix is A LOT more complicated than Elements. To be honest, in the end, it is probably faster for me to implement what you want than explaining the details. But inasmuch as I want to get you up to speed, I have other projects to attend to. But I'll see what I can do to make this a high priority. I appreciate the value of your other contributions and it's probably the least I can do.

**Comment by Xeverous** [comment]
> collaborators go figure out by themselves how things work internally (there were no internal docs). And the internals of Spirit, Fusion, and Phoenix are A LOT more complicated than Elements.

This sounds just bad. Some parts of internal logic are not documented at all and I would be afraid of losing time on questionable implementations or bugs like recent crash in sliders example. This might then cause formal maintainers to need even more time to fix such problems.

> I'll see what I can do to make this a high priority. I appreciate the value of your other contributions and it's probably the least I can do.

Yeah, I wanted to point out that I have already contributed a lot (in my free time) (especially recent file dialogs work) so I think expecting some help from you is reasonable. Initially I wanted a modern, clean GUI library and after seeing elements, just only add missing widgets for my needs (radio button, tabs, file dialogs) but I have gone much further than initially expected ... I didn't thought I would be opening 3-digit PRs/issues.

**Comment by djowel** [comment]
> > collaborators go figure out by themselves how things work internally (there were no internal docs). And the internals of Spirit, Fusion, and Phoenix are A LOT more complicated than Elements.
> 
> This sounds just bad. Some parts of internal logic are not documented at all and I would be afraid of losing time on questionable implementations or bugs like recent crash in sliders example. This might then cause formal maintainers to need even more time to fix such problems.

That is normal in the course of development, esp. before it becomes release ready. That is why some call it the "bleeding edge". Once we get into a stable state, then release 1.0 master will maintain stability, but the develop branch will still experience such things.

It's fine if you are not up to it.

**Comment by djowel** [comment]
The first cut implementation is on the "notebook" branch. It will be merged to develop once it is stable.

**Comment by Xeverous** [comment]
I have checked commits on this branch and I'm satisfied with the result.

The only thing worth to mention: the current implementation of `notebook` factory has hardcoded layout (tabs above pages, left-aligned) but given how the factory looks it should be trivial to add more variations.

**Comment by djowel** [comment]
> The only thing worth to mention: the current implementation of `notebook` factory has hardcoded layout (tabs above pages, left-aligned) but given how the factory looks it should be trivial to add more variations.

Exactly! As I always say, start simple and add more features later.

**Comment by Xeverous** [comment]
Ok, my first criticizm of the implementation: https://github.com/cycfi/elements/blob/a3a294aeb3f56361480b998c8eb116e60b40cc58/lib/include/elements/element/gallery/notebook.hpp#L45-L64

I don't like the margin. Try to make something that has a layout similar to a browser. The tabs are not in the top-left corner but some offset from it, caused by the margin in this function. IMO this just looks ugly - a wasted waste above the tabs.

My propositions:

- remove it from the function (a user can already add a margin around the notebook)
- make this margin a function parameter
- more granular API so that `link_tabs` etc is a part of public interface

**Comment by djowel** [comment]
This is not the final API of course. It's just a step towards the direction I had in mind. Feel free to suggest PRs that I can evaluate.

As a matter of fact: before we go to 1.0 is the right time to improve a lot of the APIs, esp. those in the gallery. But keep in mind that the gallery is simply a collection of pre-packaged code that the user can use as-is, or used as basis for their own purposes. That said, I am with you that the gallery should strive to be as generically useful to a broader set of use-cases.

**Comment by Xeverous** [comment]
#116

- Notebook widgets can be arbitrarily nested, so I have separated widget composition from binding widget behavior. These are 2 separate things.
- I removed the margin and added another notebook factory, for horizontal layout.
- Binding behavior is now a public API function (`link_tabs`), to let users get the notebook behavior but arrange elements as they want if they do not like 2 premade factories (`vnotebook` and `hnotebook`)

I have updated the example - you should check how it looks with `hnotebook` - there is certainly something wrong with tab lengths there but given how consistent are the factories I have made (ideal mirror of v/h and alignment functions) I suspect the issue resides how `tab` element occupies space that is given to it.

**Comment by Xeverous** [comment]
There is definitely something wrong with the `tab` element. When `htile`d, it does not occupy all the space that is given to it which causes them to look bad because each has a different length. I suspected that exactly the same problem occurs in reverse situation when they are `vtile`d when they have different heights, so I tried to edit the example and discovered 2 more problems:

- the following change breaks tab switching behavior:

```diff
diff --git a/examples/notebook/main.cpp b/examples/notebook/main.cpp
index 3e8a509..d549cb9 100755
--- a/examples/notebook/main.cpp
+++ b/examples/notebook/main.cpp
@@ -27,7 +27,7 @@ auto make_tabs(view& view_)
       )
     , tab("Dream")
     , tab("Insanity")
-    , tab("Imagination")
+    , hmin_size(50, tab("Imagination"))
    );
 }
```

- the following change doesn't compile (`tab` can only take `std::string`)

```diff
diff --git a/examples/notebook/main.cpp b/examples/notebook/main.cpp
index 3e8a509..7611658 100755
--- a/examples/notebook/main.cpp
+++ b/examples/notebook/main.cpp
@@ -27,7 +27,7 @@ auto make_tabs(view& view_)
       )
     , tab("Dream")
     , tab("Insanity")
-    , tab("Imagination")
+    , tab(hmin_size(50, label("Imagination")))
    );
 }
```

**Comment by djowel** [comment]
I really should start using tests, probably starting with layout issues like this. We can't keep on using the examples as tests. I'll think about it...

**Comment by Xeverous** [comment]
That's a good point but while I can instantly find color/sizing/stretching issues in a given UI I can't imagine how would you write all the test conditions given how complex widget composing can get, even if you implemented a sort of virtual mock view/window class.

**Comment by djowel** [comment]
> That's a good point but while I can instantly find color/sizing/stretching issues in a given UI I can't imagine how would you write all the test conditions given how complex widget composing can get, even if you implemented a sort of virtual mock view/window class.

I know it's difficult. If it was easy, I would have done it already :-) But then again, I think I can start with something simple, like what I did in Artist. Anything is better than nothing.

**Comment by Xeverous** [comment]
copy-paste mistake:

https://github.com/cycfi/elements/blob/40037b2b49fe23b842ac12e6957e516c7856a01e/lib/include/elements/element/gallery/tab.hpp#L13-L20

And for real issues with the tab:

- Currently it is a wrapper around button or label widget, it should be more generic like all other widgets in their factories
- Tab should implement / should help implementing sizing behavior that is correct for a typical notebook: if tabs are `htile`d, they should have equal height; if tabs are `vtile`d, they should have equal width.

**Comment by djowel** [comment]
> * Currently it is a wrapper around button or label widget, it should be more generic like all other widgets in their factories

The `tab_element_base` class is a gallery element. Of course, you can have different kinds of tabs as long as it follows the basic_choice, or selectable interface.

> * Tab should implement / should help implementing sizing behavior that is correct for a typical notebook: if tabs are `htile`d, they should have equal height; if tabs are `vtile`d, they should have equal width.

Not sure what you mean. Pictures will help.

**Comment by Xeverous** [comment]
> The `tab_element_base` class is a gallery element. Of course, you can have different kinds of tabs as long as it follows the basic_choice, or selectable interface.

And I think the problem lies here. Why the `tab_element_base` has a `std::string` member? It's duplicating button/label state. It should only be a behavior-defining widget, not a state widget - if someone wants a tab with text they should do `tab(label("string"))`, if they want a tab with icon they should do `tab(icon(...))` and so on. The `tab` factory function should only add required code to make tab switching work and apply different background colors from the theme depending whether the tab is currently enabled. I'm trying here to follow your library philosophy of reusable elements and no artificial/accidental API limitations.

**Comment by Xeverous** [comment]
>> [...] if tabs are htiled, they should have equal height; if tabs are vtiled, they should have equal width

> Not sure what you mean. Pictures will help.

I have marked in red space that should be occupied by tab's clickable area. I

![tab_widget](https://user-images.githubusercontent.com/20820409/83036548-05db6380-a03b-11ea-8584-e90cd45a882a.png)

**Comment by Xeverous** [comment]
Another wording: on the left, "Dream" and "Insanity" tabs should be higher, because the third tab has larger height; on the right, the "Dream" and "Insanity" tabs should be wider, because the third tab has larger width.

**Comment by djowel** [comment]
> And I think the problem lies here. Why the `tab_element_base` has a `std::string` member? It's duplicating button/label state.

Show me exactly where it is duplicated. When you are claiming things like this, it always helps to be precise.

**Comment by Xeverous** [comment]
> Show me exactly where it is duplicated

https://github.com/cycfi/elements/blob/40037b2b49fe23b842ac12e6957e516c7856a01e/lib/include/elements/element/gallery/tab.hpp#L20-L32

https://github.com/cycfi/elements/blob/40037b2b49fe23b842ac12e6957e516c7856a01e/lib/include/elements/element/misc.hpp#L150-L166

IMO the `tab` factory function should accept an arbitrary widget and store that widget and only appy necessary behavior for the notebook to function properly.

**Comment by djowel** [comment]
> Another wording: on the left, "Dream" and "Insanity" tabs should be higher, because the third tab has larger height; on the right, the "Dream" and "Insanity" tabs should be wider, because the third tab has larger width.

This looks like the same issue as you had before with tiles. That is how tiles behave. (see layout docs). I'll see if there's a generic solution, or if it is an entirely different problem altogether.

**Comment by djowel** [comment]
> IMO the `tab` factory function should accept an arbitrary widget and store that widget and only appy necessary behavior for the notebook to function properly.

Still does not make sense. Here's the *real* base class.

**Comment by djowel** [comment]
Look, I think you are misunderstanding the purpose of the gallery. It is supposed to be a collection of useful things, as generic as possible, but not try hard to be TOO generic. The generic clases are the basic elements outside the gallery from which the gallery stuff are made from.

**Comment by Xeverous** [comment]
Ok, I get what you mean but, given current state of the repository, how much code would I need to make a tab that has something different than text?

I think that elements should supply a factory function that allows to put arbitrary widget inside a notebook tab, this way then user could then do something like this:

```cpp
vnotebook(
    view,
    pages,
    tab_factory(label("foo")),
    tab_factory(label("bar")),
    tab_factory(image("...")),
    tab_factory(htile(icon_gear, label("settings")))
)
```

**Comment by djowel** [comment]
> Ok, I get what you mean but, given current state of the repository, how much code would I need to make a tab that has something different than text?

I'll leave that up to you as an exercise ;-) Make your own based on tab_element.

**Comment by djowel** [comment]
> > Ok, I get what you mean but, given current state of the repository, how much code would I need to make a tab that has something different than text?
> 
> I'll leave that up to you as an exercise ;-) Make your own based on tab_element.

Oh, and the button gallery should also have some interesting examples that you can mix and match.

**Comment by Xeverous** [comment]
> This looks like the same issue as you had before with tiles. That is how tiles behave. (see layout docs). I'll see if there's a generic solution, or if it is an entirely different problem altogether.

I think this time the issue is different.

Previously all widgets were stretchable (buttons, labels, input boxes), but I only wanted input boxes to stretch. It was fixed by adding `no_hstretch` around button and label columns, which caused them to not stretch which caused input boxes column to take all of the extra horizontal space.

This time there is just 1 row or column of tabs but nothing stretches.

**Comment by djowel** [comment]
> > This looks like the same issue as you had before with tiles. That is how tiles behave. (see layout docs). I'll see if there's a generic solution, or if it is an entirely different problem altogether.
> 
> I think this time the issue is different.
> 
> Previously all widgets were stretchable (buttons, labels, input boxes), but I only wanted input boxes to stretch. It was fixed by adding `no_hstretch` around button and label columns, which caused them to not stretch which caused input boxes column to take all of the extra horizontal space.
> 
> This time there is just 1 row or column of tabs but nothing stretches.

I think they are the same at a fundamental level. What you are describing are effects of this, but I am not sure if I can explain completely, easily.

**Comment by djowel** [comment]
> I think they are the same at a fundamental level. What you are describing are effects of this, but I am not sure if I can explain completely, easily.

Here: http://cycfi.github.io/elements/layout#vertical-tiles

Keypoints:
- Space is allocated for each child element following the child’s natural minimum vertical limit.
- If a child element’s maximum limits in either dimension is exceeded, the element will be aligned to the top-left.

**Comment by Xeverous** [comment]
> I'll leave that up to you as an exercise ;-) Make your own based on tab_element.

Ok, I will try it and see the code. Maybe then I will turn this into a PR which will express my intent clearer. I understand your points about the balance of ease of use and genericity but I don't get why the simplest "tab with text" is not actually using the `label` or `button` elements.

**Comment by djowel** [comment]
> > I'll leave that up to you as an exercise ;-) Make your own based on tab_element.
> 
> Ok, I will try it and see the code. Maybe then I will turn this into a PR which will express my intent clearer. I understand your points about the balance of ease of use and genericity but I don't get why the simplest "tab with text" is not actually using the `label` or `button` elements.

Very valid points! Let's see how it goes. The gallery **can* evolve and be more generic as needed. The button gallery is an example.

**Comment by Xeverous** [comment]
> If a child element’s maximum limits in either dimension is exceeded, the element will be aligned to the top-left.

This is the root cause then. If I have few widgets tiled vertically:

- the vertical size of the tiling is the **sum** of their minimum vertical requirements
- the horizontal size of the tiling is the **max** horizontal size among them

The problem is, how can I make other widgets to resize to that max horizontal size?

___

![image](https://user-images.githubusercontent.com/20820409/83043921-c49b8180-a043-11ea-9c18-c3fe576633d8.png)

On this example there are 2 columns. The column with buttons is intentionally wrapped in `no_hstretch` so that all of the extra space is given to input boxes. And this works because for magic reason input boxes expand to infinity. But tabs do not, they always keep their minum size. That's my problem - is there any way to make tabs auto-expand whenever there is any extra space?

**Comment by djowel** [comment]
> > If a child element’s maximum limits in either dimension is exceeded, the element will be aligned to the top-left.
> 
> This is the root cause then. If I have few widgets tiled vertically:
> 
> * the vertical size of the tiling is the **sum** of their minimum vertical requirements
> * the horizontal size of the tiling is the **max** horizontal size among them
> 
> The problem is, how can I make other widgets to resize to that max horizontal size?

Agreed. I'm thinking about it as we speak. I'll need to investigate the effects if I change this behavior to have elements use more space. The tab-element, and labels for that matter, have fixed size: https://github.com/cycfi/elements/blob/notebook/lib/src/element/gallery/tab.cpp#L47

**Comment by Xeverous** [comment]
> Update: Oh, maybe that's what you want ("any way to make tabs auto-expand whenever there is any extra space")?

Yes, that's what I want. Input boxes can resize horizontally for obvious reasons, but labels have no point in it as a fixed text has always the same size, so why not add an alignment near them? I wanted to immediately test it with `tab(align_center(label("test")))` but then the experiment is blocked by the factory expecting `std::string`. Looks like I will need to work on the tab factory PR first.

**Comment by Xeverous** [comment]
![image](https://user-images.githubusercontent.com/20820409/83070694-1eb03d00-a06c-11ea-94b2-705f70daef5e.png)

I'm getting better at it. Now I understand completely how aligning+tiling+stretching combined interact. Expect a PR for the tab interface and a modified notebook example soon.

**Comment by Xeverous** [comment]
While implementing the PR, I have identified some bugs in proxy:

https://github.com/cycfi/elements/blob/40037b2b49fe23b842ac12e6957e516c7856a01e/lib/include/elements/element/proxy.hpp#L61-L88

Both of these functions take `Subject&& subject` argument, but:

- There is no perfect forwarding because `Subject` is not deduced. It's already specified by the class template and both the constructor and `subject` setter are simply using it. Perfect forwarding is only possible when the function template can apply reference collapsing during type deduction.
- Because of how we use proxy (we always store values of non-reference type), the `std::decay` is a no-op. It has no purpose when `Subject` is not deduced and at this point (inside the class) `Subject` already must be a non-reference type.
- Beacuse `Subject&&` is not deduced, both `std::forward`s are no-op.
- Beacuse `Subject&&` is not deduced, both the constructor and the setter are expecting rvalues. This hit me during some tab factory works when I supplied an lvalue to the constructor. You can trigger the error by any lvalue object: `proxy<element>(empty())` works, but `auto e = empty(); proxy<element>(e);` fails on forbidden lvalue-to-rvalue convertion.

**Comment by djowel** [comment]
> but labels have no point in it as a fixed text has always the same size

You really should not make strong claims like that without knowing the underlying reasons ;-)

**Comment by Xeverous** [comment]
So, are there more reasons why labels always have fixed size?

**Comment by djowel** [comment]
> Both of these functions take `Subject&& subject` argument, but:
> 
> * There is no perfect forwarding because `Subject` is not deduced. It's already specified by the class template and both the constructor and `subject` setter are simply using it. Perfect forwarding is only possible when the function template can apply reference collapsing during type deduction.

Correct. Indeed. The ctor should probably take a template to allow convertibility, etc. That is the intent here IIRC.

> * Because of how we use proxy (we always store values of non-reference type), the `std::decay` is a no-op. It has no purpose when `Subject` is not deduced and at this point (inside the class) `Subject` already must be a non-reference type.

The way we use it today does not mean it will be so in the future. There might be a possibility of using references (although we already have the indirect classes), but I am not ready to stifle the interface just yet.

**Comment by djowel** [comment]
> So, are there more reasons why labels always have fixed size?

Good. It's always good to ask first before assuming something is "wrong" :) I suggest you avoid such assumptions without knowing the underlying reasons. Don't assume anything even if it seems obvious to you initially.

Consider this use-case:
Imagine this: I want an icon (text icon) at the left of some text with some *exact* margin. How do I do that if the icon and text have flexible sizes? Have a look at the icon+text buttons, for example:

https://github.com/cycfi/elements/blob/master/lib/include/elements/element/gallery/button.hpp#L85-L105

Can you imagine now how that code will be messed up with the behavior that you propose?

Moreover... Should icons, or images have fixed sized? If not, should they stretch to fit or align to center? 

Whatever your answer is, these are design and conceptual questions and there is really no right or wrong answers. It is a choice. IMO, they should be fixed and there should be consistent behavior. After all, there are always align proxies that you can use, and it is easy to add stretch-to-fit proxies if you want some elements to scale fluidly. 

That is the underlying design: 
1. Keep the elements as simple and consistent as possible. 
2. Use proxy ("decorators") to change the basic behavior.

**Comment by Xeverous** [comment]
I managed to workaround current broken forwarding in the tab PR (I will not attempt fixing proxy on this branch, is the job for develop), however I have encountered another bug:

```
lib/include/elements/element/button.hpp:90:18: error: no match for 'operator=' (operand types are 'std::array<std::shared_ptr<cycfi::elements::element>, 2>::value_type' {aka 'std::shared_ptr<cycfi::elements::element>'} and 'std::shared_ptr<cycfi::elements::tab_element<false, cycfi::elements::label_gen<cycfi::elements::basic_label_base<cycfi::elements::default_label> > > >')
   90 |       (*this)[0] = share(std::forward<W1>(off));
```

shared pointer has a converting constructor, so I was surprised why it can not upcast the pointed type to base `element` type. Everything after this error are failed `enable_if`s from different `operator=` overloads which don't tell me much apart from the fact that `std::is_assignable` said no.

So I went ahead and made a simple experiment:

```cpp
        tab_element<false, label_gen<basic_label_base<default_label>>>* p1 = nullptr;
        auto p2 = static_cast<element*>(p1);
```

```
main.cpp:20:36: error: 'cycfi::elements::element' is an inaccessible base of 'cycfi::elements::tab_element<false, cycfi::elements::label_gen<cycfi::elements::basic_label_base<cycfi::elements::default_label> > >'
   20 |  auto p2 = static_cast<element*>(p1);
      |                                    ^
```

*somewhere* across the hierarchy there is a `class` that uses implicit private inheritance - maybe a bug introduced in recent size refactorings (although there are no such types in this instantiation). Now it's just time for some `grep` hunting.

**Comment by Xeverous** [comment]
Surprise, it was actually may fault because I edited the `tab_element` incorrectly:

```
lib/include/elements/element/gallery/tab.hpp:17:   class tab_element : proxy<Subject>
```

Still, I searched the entire repository just in case - not more such mistakes were found.

**Comment by Xeverous** [comment]
Ok, please build 2070c88e7ca7b26df011b1b430c695e6d096fd88 and look at its result.

It's not a PR but it should clearly express what I had in mind.

Few notable things:

- Initially tab was a simple label-like factory accepting only `std::string`. After this commit the `tab` can accept any element.
- `tab_element` is now a proxy element which only adds a background behind the proxied element.
- I have updated the example, on this commit you can see that hnotebook looks correctly
- As a result, `draw_tab` has been significantly simplified. It's meant to be just a background that reacts with color changes depending whether tab is enabled (`state`) and/or highlighted (`hilite`). `hilite` does not work (background doesn't react to hovering) but it was broken before this commit (looks like some `refresh` call is missing somewhere)

**Comment by djowel** [comment]
Looks good to me! Now I see what you mean. Yes, this is in line with the design goals of Elements, and I am happy :-)

**Comment by Xeverous** [comment]
First you educated me how to implement progress bar element correctly to be generic. Now I'm improving the tab because I see a mistake like you pointed me then.

Before I make this change an actual PR, I will try to repair proxy. You can see how the tab currently uses a `copy` function to workaround (unwanted) rvalue requirement.

**Comment by djowel** [comment]
> First you educated me how to implement progress bar element correctly to be generic. Now I'm improving the tab because I see a mistake like you pointed me then.

Then you educated me to be consistent with the design philosophies :) I'm glad you get it. Sometimes I forget too :-) Thanks for your patience in reminding me :-)

**Comment by Xeverous** [comment]
#122, likely CI problem

**Comment by Xeverous** [comment]
Since #122 is merged, I think this branch will be ready to be merged onto develop when the copying interface is implemented.

**Comment by Xeverous** [comment]
#128 brings tab in line with layered button and other button factories:

- do not copy any elements
- ask caller for 2 elements explicitly

**Comment by Xeverous** [comment]
Can this be merged to develop now? The only thing it still needs is the state duplication refactor, just like some buttons.

**Comment by djowel** [comment]
See branch: https://github.com/cycfi/elements/tree/use-basic_button

Which I branched off "notebook". Check out the changes. This is what I had in mind with the basic_button vs. layered_button, which might become obsolete because that too can be a specialization of basic_button.

**Comment by Xeverous** [comment]
Great, I will check the diff vs notebook branch.

**Comment by Xeverous** [comment]
Yes, this is what I had in mind. Go for it on every element that creates duplicate elements in its factory.

I have 2 goals:

- Being able to create a simple element with good default behavior (eg `tab("a string")` which reacts to hover and selection using theme "hover-highlight" and "selected" colors).
- Being able to create a custom element that will render differently based on the state, eg a tab with text/icon/image/canvas and custom render on hower/chosed states. This is possible already thanks to `receiver<bool>` and `sender<bool>` interfaces, I'm just looking if the library can ease creating such widgets even more.

And btw, on this branch the notebook example is broken. Tabs can only be switched on once and stay "on" forever.

I'm encouraging you to merge notebook or even use-basic_button branche to develop, as my other work (my own project, bug fixes, theme editor) are depending on more and more bug fixes/elements.

**Comment by djowel** [comment]
> And btw, on this branch the notebook example is broken. Tabs can only be switched on once and stay "on" forever.

Fixed. Simplified traversal code along the way.

> I'm encouraging you to merge notebook or even use-basic_button branche to develop, as my other work (my own project, bug fixes, theme editor) are depending on more and more bug fixes/elements.

Merged into develop. Still open to refinements such as aesthetics, API, etc.

**Comment by Xeverous** [comment]
Great. There is still some small weirdness in the notebook example - the first tab text is highlighted whenever mouse leaves the window.

**Comment by djowel** [comment]
> Great. There is still some small weirdness in the notebook example - the first tab text is highlighted whenever mouse leaves the window.

I'll look into it.

**Comment by djowel** [comment]
> I'll look into it.

Fixed. I'll close this now.
